core/services/workflows/syncer/workflow_registry.go

```diff
@@ -353,13 +398,207 @@ func (w *workflowRegistry) readRegistryEvents(ctx context.Context, don capabilit
 			}
 
 			for _, event := range events {
-				err := w.handler.Handle(ctx, event.Event)
+				select {
+				case <-ctx.Done():
+					w.lggr.Debug("readRegistryEventsLoop stopped during processing")
+					return
+				case w.eventCh <- event.Event:
+				}
+			}
+		}
+	}
+}
+
+// syncUsingEventStrategy syncs workflow registry contract state by watching for events on the contract.
+// It first loads the workflow metadata from the contract state as WorkflowRegistered events,
+// and then starts a goroutine with one loop for handling the events.
+func (w *workflowRegistry) syncUsingEventStrategy(ctx context.Context, don capabilities.DON, reader ContractReader) {
+	w.lggr.Debugw("Loading initial workflows for DON", "DON", don.ID)
+
+	workflowMetadata, loadWorkflowsHead, err := w.getWorkflowMetadata(ctx, don, reader)
+	// TODO: if this fails, we should still start the event looping
+	if err != nil {
+		w.lggr.Errorw("failed to load initial workflows", "err", err)
+		return
+	}
+
+	w.lggr.Debugw("Rehydrating existing workflows", "len", len(workflowMetadata))
+	for _, workflow := range workflowMetadata {
+		select {
+		case <-ctx.Done():
+			w.lggr.Debug("shut down during initial workflow registration")
+			return
+		case w.eventCh <- workflowAsEvent{
+			Data: WorkflowRegistryWorkflowRegisteredV1{
+				WorkflowID:    workflow.WorkflowID,
+				WorkflowOwner: workflow.Owner,
+				DonID:         workflow.DonID,
+				Status:        workflow.Status,
+				WorkflowName:  workflow.WorkflowName,
+				BinaryURL:     workflow.BinaryURL,
+				ConfigURL:     workflow.ConfigURL,
+				SecretsURL:    workflow.SecretsURL,
+			},
+			EventType: WorkflowRegisteredEvent,
+		}:
+		}
+	}
+
+	// Poll for all workflow related events
+	ets := []WorkflowRegistryEventType{
+		ForceUpdateSecretsEvent,
+		WorkflowActivatedEvent,
+		WorkflowDeletedEvent,
+		WorkflowPausedEvent,
+		WorkflowRegisteredEvent,
+		WorkflowUpdatedEvent,
+	}
+
+	w.readRegistryEventsLoop(ctx, ets, don, reader, loadWorkflowsHead.Height)
+}
+
+// workflowMetadataToEvents compares the workflow registry workflow metadata state against the engine registry's state.
+// Differences are handled by the event handler by creating events that are sent to the events channel for handling.
+func (w *workflowRegistry) workflowMetadataToEvents(ctx context.Context, workflowMetadata []GetWorkflowMetadata, donID uint32) []workflowAsEvent {
+	var events []workflowAsEvent
+
+	// Keep track of which of the engines in the engineRegistry have been touched
+	seenMap := map[string]bool{}
+
+	for _, wfMeta := range workflowMetadata {
+		// TODO: ensure that the WorkflowRegisteredEvent sets the engine registry as the very last step
+		// TODO: ensure that the WorkflowDeletedEvent clears the engine registry as the very last step
```

Comment:
leaving these unaddressed?

Reply:
Handled in https://github.com/smartcontractkit/chainlink/pull/17290